app/app.go

```diff
@@ -329,17 +329,22 @@ func New(
 	)
 	// transfer stack contains (from top to bottom):
 	// - Token Filter
+	// - Packet Forwarding Middleware
 	// - Transfer
 	var transferStack ibcporttypes.IBCModule
 	transferStack = transfer.NewIBCModule(app.TransferKeeper)
-	transferStack = tokenfilter.NewIBCMiddleware(transferStack)
-	transferStack = packetforward.NewIBCMiddleware(
+	packetForwardMiddleware := packetforward.NewIBCMiddleware(
 		transferStack,
 		app.PacketForwardKeeper,
 		0, // retries on timeout
 		packetforwardkeeper.DefaultForwardTransferPacketTimeoutTimestamp, // forward timeout
 		packetforwardkeeper.DefaultRefundTransferPacketTimeoutTimestamp,  // refund timeout
 	)
+	// packetForwardMiddleware is used only for version 2
+	transferStack = module.NewVersionedIBCModule(packetForwardMiddleware, transferStack, v2, v2)
+	// token filter wraps packer forward middleware and is thus the first module in the transfer stack
+	tokenFilterMiddelware := tokenfilter.NewIBCMiddleware(transferStack)
```

Comment:
[question] is it possible to version middleware or only the surrounding IBC modules? The line below this:

```
	transferStack = module.NewVersionedIBCModule(tokenFilterMiddelware, transferStack, v1, v2)
```

is kinda difficult to follow b/c IIUC it composed the previous transfer stack that contained PFM (v2 to v2) into a new transfer stack that is (v1 to v2). Just an idea but I wonder if it would it be easier to follow if we versioned middleware instead of the surrounding IBC module. IDK what that would look like yet so feel free to ignore.

Reply:
This is how middleware works though: it's essentially modules that keep wrapping one another (similar to the antehandler design)

Reply:
Makes sense. Resolving.